### PR TITLE
feat: add on validation end callback for EstimatorTrial [DET-4537]

### PR DESCRIPTION
**Pull request by aaron276h**
## Description
This feature allow users to keep track if their next checkpoint is their best checkpoint so far.

## Test Plan
Added unit test coverage.

### CODE REVIEW

File: harness/determined/estimator/_callback.py [hunk 1]
```diff
@@ -29,6 +31,16 @@ def on_checkpoint_end(self, checkpoint_dir: str) -> None:
         """
         pass
 
+    def on_validation_end(self, validation_metrics: Dict[str, Any]) -> None:
+        """
+        Run after every validation..
+
+        .. warning::
+            If distributed or parallel training is enabled, this callback is executed
+            only on the chief GPU (rank = 0).
```
Comment (rb-determined-ai) [inline harness/determined/estimator/_callback.py:40]: Why do we impose this limitation? Wouldn't this be a natural place for e.g. adjusting the optimizer learning rate?

**Review by rb-determined-ai — COMMENTED** [review]
> This feature allow users to keep track if their next checkpoint is their best checkpoint so far

How would someone actually track this?

**Comment by rb-determined-ai** [comment]
Overall my take here is that we already support on-validation-end behavior using RunHooks from the EvalSpec, and this seems redundant.

**Comment by aaron276h** [comment]
@rb-determined-ai yeah I agree, this PR doesn't address the underlying feature request that was made so I am going to close it for now.
